### PR TITLE
Update Regex source generator to use new GetBestTypeByMetadataName method

### PR DESCRIPTION
**Pull request by stephentoub**
cc: @steveharter

### CODE REVIEW

**Review by danmoseley — APPROVED** [review]
can we add a test/ extend an existing test?

**Comment by stephentoub** [comment]
> can we add a test/ extend an existing test?

I added one.

File: src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs [hunk 1]
```diff
@@ -513,6 +534,23 @@ public partial struct E
...
 
+        private static byte[] CreateAssemblyImage(string source, string assemblyName)
+        {
+            CSharpCompilation compilation = CSharpCompilation.Create(
+                assemblyName,
+                new[] { CSharpSyntaxTree.ParseText(source, CSharpParseOptions.Default.WithLanguageVersion(LanguageVersion.Preview)) },
+                s_refs.ToArray(),
+                new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            var ms = new MemoryStream();
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:545]: `using`? not that we are consistent elsewhere in tests.

**Review by danmoseley — APPROVED** [review]

File: src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:545]: Disposing a memory stream is a nop.

Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:545]: Yes but I assumed it avoids finalizer pressure. Not that it matters here.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs:545]: There's no finalizer.

**Review by eerhardt — APPROVED** [review]
:shipit:

**Comment by stephentoub** [comment]
Thanks, @eerhardt. I opened an issue to track it and added a commit to suppress the tests.

**Pull request merged**